auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -36,73 +46,235 @@
             txMetaDataStore = (KeyValueStore<JsonObject, JsonObject>) context.getStateStore(STORE_NAME);
+
+            context.schedule(Duration.ofSeconds(1), PunctuationType.WALL_CLOCK_TIME, ts -> {
+                enrichAndEmitBufferedEvents();
+            });
+
         }
 
         @Override
         public KeyValue<JsonObject, JsonObject> transform(JsonObject key, JsonObject value) {
+            boolean enrichedAllBufferedEvents = enrichAndEmitBufferedEvents();
+
+            if (!enrichedAllBufferedEvents) {
+                bufferChangeEvent(key, value);
+                return null;
+            }
+
+            KeyValue<JsonObject, JsonObject> enriched = enrichWithTxMetaData(key, value);
+            if (enriched == null) {
+                bufferChangeEvent(key, value);
+            }
+
+            return enriched;
+        }
+
+        /**
+         * Enriches the buffered change event(s) with the metadata from the associated
+         * transactions and forwards them.
+         *
+         * @return {@code true}, if all buffered events were enriched and forwarded,
+         *         {@code false} otherwise.
+         */
+        private boolean enrichAndEmitBufferedEvents() {
+            Optional<BufferOffsets> seq = bufferOffsets();
+
+            if (!seq.isPresent()) {
+                return true;
+            }
+
+            BufferOffsets sequence = seq.get();
+
+            boolean enrichedAllBuffered = true;
+
+            for(long i = sequence.getFirstValue(); i < sequence.getNextValue(); i++) {
+                JsonObject buffered = streamBuffer.get(i);
+
+                LOG.info("Processing buffered change event for key {}", buffered.getJsonObject("key"));
+
+                KeyValue<JsonObject, JsonObject> enriched = enrichWithTxMetaData(buffered.getJsonObject("key"), buffered.getJsonObject("changeEvent"));
+                if (enriched == null) {
+                    enrichedAllBuffered = false;
+                    break;
+                }
+
+                context.forward(enriched.key, enriched.value);
+                streamBuffer.delete(i);
+                sequence.incrementFirstValue();
+            }
+
+            streamBuffer.put(BUFFER_OFFSETS_KEY, sequence.toJson());
+
+            return enrichedAllBuffered;
+        }
+
+
+        /**
+         * Adds the given change event to the stream-side buffer.
+         */
+        private void bufferChangeEvent(JsonObject key, JsonObject changeEvent) {
+            LOG.info("Buffering change event for key {}", key);
+
+            BufferOffsets sequence = bufferOffsets().orElseGet(BufferOffsets::initial);
+
+            JsonObject wrapper = Json.createObjectBuilder()
+                    .add("key", key)
+                    .add("changeEvent", changeEvent)
+                    .build();
+            streamBuffer.put(sequence.getNextValueAndIncrement(), wrapper);
```

Comment:
You might want to "batch" these two consecutive put-calls together. You simply use putAll Method for that and hand it a List of both entries. The underlying RocksDB store applies the batch atomically then. WDYT?

Reply:
Good idea. I'll give it a try (sorry for the late reply, lots of travel lately...)

Reply:
Done.